src/main/scala/analysis/DSA.scala

```diff
@@ -0,0 +1,221 @@
+ * local, bottom-up, top-down results in member variables locals, bu and td respectively.
+ * @param program program to be analysed
+ * @param symResults result of symbolic access analysis
+ * @param constProp
+ * @param globals
+ * @param globalOffsets
+ * @param externalFunctions
+ * @param reachingDefs
+ * @param writesTo mapping from procedures to registers they change
+ * @param params mapping from procedures to their parameters
+ */
+class DSA(program: Program,
+            symResults: Map[CFGPosition, Map[SymbolicAccess, TwoElement]],
+            constProp: Map[CFGPosition, Map[Variable, FlatElement[BitVecLiteral]]],
+            globals: Set[SymbolTableEntry], globalOffsets: Map[BigInt, BigInt],
+            externalFunctions: Set[ExternalFunction],
+            reachingDefs: Map[CFGPosition, Map[Variable, Set[CFGPosition]]],
+            writesTo: Map[Procedure, Set[Register]],
+            params:  Map[Procedure, Set[Variable]]
+         ) extends Analysis[Map[Procedure, DSG]] {
+
+  val locals : mutable.Map[Procedure, DSG] = mutable.Map()
+  val bu: mutable.Map[Procedure, DSG] = mutable.Map()
+  val td: mutable.Map[Procedure, DSG] = mutable.Map()
+
+  val stackPointer = Register("R31", 64)
+  val returnPointer = Register("R30", 64)
+  val framePointer = Register("R29", 64)
+
+  val ignoreRegisters: Set[Variable] = Set(stackPointer, returnPointer, framePointer)
+
+  def findLeaf(procedure: Procedure): Set[Procedure] =
+    if CallGraph.succ(procedure).isEmpty then
+      Set(procedure)
+    else
+      CallGraph.succ(procedure).foldLeft(Set[Procedure]()){
+        (s, proc) => s ++ findLeaf(proc)
+      }
+
+
+  var visited = Set[Procedure]()
+  val queue = mutable.Queue[Procedure]()
+
+  override def analyze(): Map[Procedure, DSG] = {
+    var domain : Set[Procedure] = Set(program.mainProcedure)
+    val stack : mutable.Stack[Procedure] = mutable.Stack()
+    stack.pushAll(program.mainProcedure.calls)
+
+    // calculate the procedures used in the program
+    while stack.nonEmpty do
+      val current = stack.pop()
+      domain += current
+      stack.pushAll(current.calls.diff(domain))
+
+
+    // perform local analysis on all procs
+    domain.foreach(
+      proc =>
+        val dsg = Local(proc, symResults, constProp, globals, globalOffsets, externalFunctions, reachingDefs, writesTo, params).analyze()
+
+        locals.update(proc, dsg)
+        bu.update(proc, dsg.cloneSelf())
+    )
+
+    val leafNodes = findLeaf(program.mainProcedure)
+
+    leafNodes.foreach(
+      proc =>
+        assert(locals(proc).callsites.isEmpty)
+        visited += proc
+        val preds : Set[Procedure] = CallGraph.pred(proc)
+        queue.enqueueAll(CallGraph.pred(proc).diff(visited).intersect(domain))
+    )
+
+    // bottom up phase
+    while queue.nonEmpty do
+      val proc = queue.dequeue()
+      visited += proc
+      queue.enqueueAll(CallGraph.pred(proc).diff(visited))
+      val buGraph = bu(proc)
```

Comment:
NoSuchElementException gets thrown here for the tests correct/indirect_call/clang_O2:GTIRB and correct/malloc_with_local3/clang_O2:GTIRB when run with the analysis on.

Reply:
this was due to inlined procedures being added to the queue for analysis. it's now resolved